wstr/src/ops.rs

```diff
@@ -462,6 +462,47 @@ impl<'a> WStrToUtf8<'a> {
         }
     }
 
+    /// Map the given UTF-16 code unit index to its corresponding UTF-8 code unit index.
+    pub fn utf8_index(&self, utf16_index: usize) -> Option<usize> {
+        self.translate_index(utf16_index, false)
+            .map(|(utf8_index, _)| utf8_index)
+    }
+
+    /// Map the given UTF-8 code unit index to its corresponding UTF-16 code unit index.
+    pub fn utf16_index(&self, utf8_index: usize) -> Option<usize> {
+        self.translate_index(utf8_index, true)
+            .map(|(_, utf16_index)| utf16_index)
+    }
+
+    fn translate_index(&self, index: usize, is_utf8: bool) -> Option<(usize, usize)> {
+        let ascii_prefix_len = self.head.len();
+        if index < ascii_prefix_len {
+            return Some((index, index));
+        }
+
+        if self.tail.is_empty() {
+            return None;
+        }
+
+        let mut utf8_tail_pos = 0;
+        let mut utf16_tail_pos = 0;
+
+        while if is_utf8 {
+            utf8_tail_pos + ascii_prefix_len < index
+        } else {
+            utf16_tail_pos + ascii_prefix_len < index
+        } {
+            let c = self.tail[utf16_tail_pos..].chars().next()?.ok()?;
+            utf8_tail_pos += c.len_utf8();
+            utf16_tail_pos += c.len_utf16();
+        }
+
+        Some((
+            ascii_prefix_len + utf8_tail_pos,
+            ascii_prefix_len + utf16_tail_pos,
+        ))
+    }
+
```

Comment:
Note that similar code already exists in https://github.com/ruffle-rs/ruffle/blob/master/core/src/avm2/regexp.rs#L380.

Unfortunately, sharing code between this and `avm2/regexp.rs` isn't trivial as the ownership is different (a simple reference here, and an owned `AvmString<'gc>` in `regexp.rs`), but maybe there's a clever way of doing it?

Reply:
That's right, sharing the code is not trivial, I thought about this.

Basically (1) I could just use `WStrToUtf8::new(&self.text).utf{8,16}_index(utf{16,8}_index)` when the indices are looked up, but that's not a perfect solution because it would force `O(n^2)`/`O(n^2)`/`O(n)` forward-pass/backward-pass/lookup complexity instead of the current `O(n)`/`O(n^2)`/`O(n)` and potential `O(n)`/`O(n)`/`O(1)+`. That's however the simplest solution which is not affected by the ownership problem.

If I decide to have state (which is required for anything better than `O(n)` lookup) I have the ownership problem. If I understand gc-arena well enough, `Collect` implementation is required to keep track of `Gc` references. The `derive(Collect)` automatically implements it hierarchically, but every non `'static` field also has to implement `Collect`. In this case I cannot implement `Collect` for `WStrToUtf8`, because (2) from the point of view of wstr it makes no sense, and (3) in regexp I have the E0117 problem (assuming that implementing `Collect` for `WStrToUtf8` is even possible at this point). I could make it `'static` by (4) not referencing head and tail but passing the string for every lookup, but that would be difficult to use, unsafe, and unnatural. I could (5) make it owned so that there is no pointer. That would require creating something like `WStrToUtf8Owned` and converting between these two types. There is also an issue of `WStrToUtf8` being a struct of pointers so implementing `ToOwned`/`Borrow` would be difficult? `WStrToUtf8Owned` would have to be implemented in wstr, as otherwise I would have to make the head/tail implementation detail public. I cannot replace `WStrToUtf8` with `WStrToUtf8Owned` as it would create owned instances every time a conversion to UTF-8 is done. I could also (6) implement `Collect` for `CachedText` manually, but that would require an unsafe implementation and I think more knowledge about gc-arena than I have.

It seems that the option (5) is the best, what do you think? Please point any mistakes or shortcomings in my analysis above — I'm a Rust beginner :)

Reply:
> In this case I cannot implement Collect for WStrToUtf8
I'm not exactly sure of the context, but if `WStrToUtf8` is a field on a garbage-collected `Collect`-implementing object, you can mark the field as `#[collect(requires_static)]`, so you don't need to implement `Collect` for `WStrToUtf8` itself, it'll just be static, even if it references other pointers.

Reply:
According to my understanding and testing for a field of type `WStrToUtf8<'a>` (it requires a lifetime) to be marked as `#[collect(require_static)]` the `'a` lifetime would have to outlive `'static`, which is not the case here. That's why I suggested making an owned version which does not require a lifetime. But I may be approaching it from a wrong direction